### PR TITLE
Halberstam nn

### PR DESCRIPTION
**Pull request by elkyhalberstam**

### CODE REVIEW

File: src/main/java/smartsnake/Brain.java [hunk 1]
```diff
@@ -1,10 +1,12 @@
...
 
+import java.io.IOException;
+
 /**
  * An interface that decides which Direction the Snake should move in based on the Snake, Garden and Food.
  */
 public interface Brain {
 
-    Direction move(Snake snake, Food food, Garden garden);
+    Direction move(Snake snake, Food food, Garden garden) throws IOException;
```
Comment (andrewoid) [inline src/main/java/smartsnake/Brain.java:10]: No.

File: src/main/java/smartsnake/Garden.java [hunk 2]
```diff
@@ -88,7 +89,12 @@ public boolean tick() {
             }
 
             Brain brain = snake.getBrain();
-            Direction newDirection = brain.move(snake, food, this);
+            Direction newDirection = null;
+            try {
+                newDirection = brain.move(snake, food, this);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
```
Comment (andrewoid) [inline src/main/java/smartsnake/Garden.java:97]: No.

File: src/main/java/smartsnake/astar/halberstam/AstarBrain.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+            neighbors.add(up);
+            neighbors.add(down);
+            neighbors.add(right);
+            neighbors.add(left);
+
+
+            for (int i = 0; i < neighbors.size(); i++) {
+                Node neighbor = neighbors.get(i);
+                if(!garden.contains(neighbor.getLocation())) {
```
Comment (andrewoid) [inline src/main/java/smartsnake/astar/halberstam/AstarBrain.java]: You can combine this if statement with the next one.

File: src/main/java/smartsnake/Brain.java [hunk 4]
```diff
@@ -1,5 +1,7 @@
 package smartsnake;
 
+import java.io.IOException;
```
Comment (andrewoid) [inline src/main/java/smartsnake/Brain.java:3]: remove

File: src/main/java/smartsnake/astar/halberstam/AstarBrain.java (hunk 3 above)
Comment (elkyhalberstam) [reply src/main/java/smartsnake/astar/halberstam/AstarBrain.java]: when I combined them I my test returned null so I left it

File: src/main/java/smartsnake/Garden.java [hunk 5]
```diff
@@ -88,7 +89,10 @@ public boolean tick() {
             }
 
             Brain brain = snake.getBrain();
-            Direction newDirection = brain.move(snake, food, this);
+            Direction newDirection = null;
+
+            newDirection = brain.move(snake, food, this);
+
```
Comment (andrewoid) [inline src/main/java/smartsnake/Garden.java:95]: remove all changes to this class

File: src/main/java/smartsnake/astar/halberstam/AstarBrain.java (hunk 3 above)
Comment (andrewoid) [reply src/main/java/smartsnake/astar/halberstam/AstarBrain.java]: Why is it suddenly null?
Figure that out.

File: src/main/java/smartsnake/neuralnetwork/halberstam/LearnAndSave.java
Comment (elkyhalberstam) [inline src/main/java/smartsnake/neuralnetwork/halberstam/LearnAndSave.java:1]: This works for 50 but it went on for over an hour when I did 50000, so there must be something wrong but I can't find out where.
Do you have any pointers?

File: src/main/java/smartsnake/neuralnetwork/halberstam/NeuralNetworkBrain.java
Comment (elkyhalberstam) [inline src/main/java/smartsnake/neuralnetwork/halberstam/NeuralNetworkBrain.java:1]: My snake ran into the wall. Do you know Where I may have gone wrong?

Comment (andrewoid) [reply src/main/java/smartsnake/neuralnetwork/halberstam/NeuralNetworkBrain.java:1]: 50 games isn't enough training. That's why I said 50_000

File: src/main/java/smartsnake/neuralnetwork/halberstam/NeuralNetworkBrain.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package smartsnake.neuralnetwork.halberstam;
+
+import basicneuralnetwork.NeuralNetwork;
+import smartsnake.*;
+import smartsnake.neuralnetwork.NeuralNetworkDataFactory;
+
+import java.io.IOException;
+
+public class NeuralNetworkBrain implements Brain {
+    NeuralNetwork neuralNetwork = new NeuralNetwork(5, 256, 4);
```
Comment (andrewoid) [inline src/main/java/smartsnake/neuralnetwork/halberstam/NeuralNetworkBrain.java:10]: Don't create the NN if you're just going to replace it with the file later.

File: src/main/java/smartsnake/neuralnetwork/halberstam/NeuralNetworkBrain.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    NeuralNetworkDataFactory dataFactory = new NeuralNetworkDataFactory();
+
+    @Override
+    public Direction move(Snake snake, Food food, Garden garden) {
+
+        try {
+            neuralNetwork = NeuralNetwork.readFromFile();
```
Comment (andrewoid) [inline src/main/java/smartsnake/neuralnetwork/halberstam/NeuralNetworkBrain.java:17]: This should be in the constructor

File: src/main/java/smartsnake/neuralnetwork/halberstam/NeuralNetworkBrain.java
Comment (elkyhalberstam) [reply src/main/java/smartsnake/neuralnetwork/halberstam/NeuralNetworkBrain.java:1]: right but when I did 50,000 it ran for over an hour so you said I must have done something wrong but I can't figure it out.
Can you point me to where my mistake might be?

Comment (andrewoid) [reply src/main/java/smartsnake/neuralnetwork/halberstam/NeuralNetworkBrain.java:1]: I don't see anything now. Let's talk tomorrow

File: src/main/java/smartsnake/astar/halberstam/AstarBrain.java (hunk 3 above)
Comment (elkyhalberstam) [reply src/main/java/smartsnake/astar/halberstam/AstarBrain.java]: I fixed this

File: src/main/java/smartsnake/neuralnetwork/halberstam/LearnAndSave.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+            boolean tick = true;
+            while (tick) {
+                double[] gardenInput = dataFactory.toInput(garden);
+                tick = garden.tick();
+                double[] output = dataFactory.toOutput(snake.getDirection());
+                neuralNetwork.train(gardenInput, output);
+                if (!tick) //because when game ends tick is false
+                {
+                    break;
+                }
```
Comment (andrewoid) [inline src/main/java/smartsnake/neuralnetwork/halberstam/LearnAndSave.java:38]: This is unnecessary.

File: src/main/java/smartsnake/neuralnetwork/halberstam/LearnAndSave.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+        NeuralNetwork neuralNetwork;
+        try {
+            neuralNetwork = NeuralNetwork.readFromFile();
+        } catch (IOException e) {
+            e.printStackTrace();
+            neuralNetwork = new NeuralNetwork(5, 256, 4);
+        }
+        NeuralNetworkDataFactory dataFactory = new NeuralNetworkDataFactory();
+
+        for (int i = 0; i < 10000; i++)
```
Comment (andrewoid) [inline src/main/java/smartsnake/neuralnetwork/halberstam/LearnAndSave.java:22]: `50_000`

**Review by andrewoid — APPROVED** [review]

**Pull request merged**